satellite/gracefulexit/endpoint.go

```diff
@@ -0,0 +1,443 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package gracefulexit
+
+import (
+	"context"
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+
+	"storj.io/storj/pkg/identity"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/rpc/rpcstatus"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/satellite/metainfo"
+	"storj.io/storj/satellite/orders"
+	"storj.io/storj/satellite/overlay"
+	"storj.io/storj/uplink/eestream"
+)
+
+// millis for the transfer queue building ticker
+const buildQueueMillis = 100
+
+// drpcEndpoint wraps streaming methods so that they can be used with drpc
+type drpcEndpoint struct{ *Endpoint }
+
+// processStream is the minimum interface required to process requests.
+type processStream interface {
+	Context() context.Context
+	Send(*pb.SatelliteMessage) error
+	Recv() (*pb.StorageNodeMessage, error)
+}
+
+// Endpoint for handling the transfer of pieces for Graceful Exit.
+type Endpoint struct {
+	log       *zap.Logger
+	interval  time.Duration
+	db        DB
+	overlaydb overlay.DB
+	overlay   *overlay.Service
+	metainfo  *metainfo.Service
+	orders    *orders.Service
+	config    Config
+}
+
+type pendingTransfer struct {
+	path             []byte
+	pieceSize        int64
+	satelliteMessage *pb.SatelliteMessage
+}
+
+// pendingMap for managing concurrent access to the pending transfer map.
+type pendingMap struct {
+	mu   sync.RWMutex
```

Comment:
Should the mutex be a pointer here?

Reply:
I don't think so. But I could be wrong.

Reply:
You are right. We don't need it to be a pointer here since we are passing the pointer of pendingMap around